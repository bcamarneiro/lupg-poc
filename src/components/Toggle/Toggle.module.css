.wrapper {
  @apply flex flex-row gap-4 items-center justify-center border p-1 rounded-xl;
}

.toggleHandle {
  @apply flex flex-row items-center justify-center  gap-2 relative px-4 py-1 text-levelup-500 rounded-xl cursor-pointer;
}
