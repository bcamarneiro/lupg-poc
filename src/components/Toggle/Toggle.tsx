import clsx from 'clsx';
import styles from './Toggle.module.css';
import { useState } from 'react';

type ToggleOption = {
  label: string;
  icon: (props?: { inverted?: boolean }) => React.ReactNode;
  value: string;
};
interface ToggleProps {
  options: ToggleOption[];
  onChange: (value: string) => void;
}

function Toggle({ options, onChange }: ToggleProps) {
  const [active, setActive] = useState(options[0].value);

  return (
    <div className={styles.wrapper}>
      {options.map((option) => (
        <div
          key={option.value}
          className={clsx(
            styles.toggleHandle,
            active === option.value ? 'bg-levelup-200 dark:bg-levelup-600' : 'bg-levelup-600 dark:bg-levelup-200',
          )}
          onClick={() => {
            setActive(option.value);
            onChange(option.value);
          }}
        >
          {option.icon({ inverted: active === option.value })}
          <span
            className={clsx(
              active === option.value
                ? 'text-levelup-600 dark:text-levelup-50'
                : 'text-levelup-200 dark:text-levelup-600',
              'leading-[1.875rem]',
            )}
          >
            {option.label}
          </span>
        </div>
      ))}
    </div>
  );
}

export default Toggle;
