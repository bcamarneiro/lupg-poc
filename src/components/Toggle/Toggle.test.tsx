import { render, screen, fireEvent } from '@testing-library/react';
import Toggle from './Toggle';

describe('Toggle component', () => {
  const mockOptions = [
    { label: 'Option 1', icon: () => <span>Icon1</span>, value: 'option1' },
    { label: 'Option 2', icon: () => <span>Icon2</span>, value: 'option2' },
  ];

  it('renders toggle options correctly', () => {
    const mockOnChange = vi.fn();
    const { asFragment } = render(
      <Toggle
        options={mockOptions}
        onChange={mockOnChange}
      />,
    );

    expect(screen.getByText('Option 1')).toBeInTheDocument();
    expect(screen.getByText('Option 2')).toBeInTheDocument();
    expect(screen.getByText('Icon1')).toBeInTheDocument();
    expect(screen.getByText('Icon2')).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onChange handler when an option is clicked', () => {
    const mockOnChange = vi.fn();
    render(
      <Toggle
        options={mockOptions}
        onChange={mockOnChange}
      />,
    );

    const option2 = screen.getByText('Option 2');
    fireEvent.click(option2);

    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith('option2');
  });

  it('updates active state when an option is clicked', () => {
    const mockOnChange = vi.fn();
    render(
      <Toggle
        options={mockOptions}
        onChange={mockOnChange}
      />,
    );

    const option1 = screen.getByText('Option 1');
    const option2 = screen.getByText('Option 2');

    expect(option1.parentElement).toHaveClass('bg-levelup-200');
    expect(option2.parentElement).toHaveClass('bg-levelup-600');

    fireEvent.click(option2);

    expect(option1.parentElement).toHaveClass('bg-levelup-600');
    expect(option2.parentElement).toHaveClass('bg-levelup-200');
  });
});
