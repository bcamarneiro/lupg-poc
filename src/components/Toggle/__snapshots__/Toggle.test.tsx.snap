// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toggle component > renders toggle options correctly 1`] = `
<DocumentFragment>
  <div
    class="_wrapper_eb77c6 border-levelup-200 dark:border-levelup-350 bg-levelup-600 dark:bg-levelup-200"
  >
    <div
      class="_toggleHandle_eb77c6 bg-levelup-200 dark:bg-levelup-600"
    >
      <span>
        Icon1
      </span>
      <span
        class="text-levelup-600 dark:text-levelup-50 leading-[1.875rem] hidden sm:inline-block"
      >
        Option 1
      </span>
    </div>
    <div
      class="_toggleHandle_eb77c6 bg-levelup-600 dark:bg-levelup-200"
    >
      <span>
        Icon2
      </span>
      <span
        class="text-levelup-200 dark:text-levelup-600 leading-[1.875rem] hidden sm:inline-block"
      >
        Option 2
      </span>
    </div>
  </div>
</DocumentFragment>
`;
